luarules/gadgets/cmd_nano_range_check.lua

```diff
@@ -0,0 +1,156 @@
+function gadget:GetInfo()
+	return {
+		name = "Construction Turrets Range Check",
+		desc = "Stops construction turrets from getting stuck on orders out of reach.",
+		author = "Nehroz",
+		date = "2024.12.01",
+		layer = 0,
+		enabled = true,
+	}
+end
+
+if not gadgetHandler:IsSyncedCode() then return end
+
+local CMD_STOP = CMD.STOP
+local gMaxUnits = Game.maxUnits
+local updateInterval = Game.gameSpeed / 2
+
+local spValidUnitID = Spring.ValidUnitID
+local spValidFeatureID = Spring.ValidFeatureID
+local spGetUnitDefID = Spring.GetUnitDefID
+local spGetUnitSeparation = Spring.GetUnitSeparation
+local spGetUnitFeatureSeparation = Spring.GetUnitFeatureSeparation
+local spGetUnitCommands = Spring.GetUnitCommands
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+
+--- Key is the nano's ID, followed by a set-like with the unitID of target currently inside the nano's range.
+--- Used for CommandArray manipulation
+---@type {[integer] : {integer : {cmdTag : integer, status : integer}}}
+local trackingTable = {}
+local tagUpdateFrameMax = 1 -- Can be bigger than 1 if you want game freezes.
+
+local statuses = {
+	outOfRange = 0,
+	inRange = 1,
+}
+
+local mobileUnits = {}
+local constructionTurretsDefs = {}
+
+for unitDefID, unitDef in ipairs(UnitDefs) do
+	if unitDef.isFactory == false and unitDef.isStaticBuilder then
+		constructionTurretsDefs[unitDefID] = {
+			maxBuildDistance = unitDef.buildDistance + unitDef.radius,
+			buildRange3D = unitDef.buildRange3D,
+		}
+	end
+
+	if unitDef.canMove then
+		mobileUnits[unitDefID] = true
+	end
+end
+
+function gadget:Initialize()
+	gadgetHandler:RegisterAllowCommand(CMD.REPAIR)
+	gadgetHandler:RegisterAllowCommand(CMD.GUARD)
+	gadgetHandler:RegisterAllowCommand(CMD.RECLAIM)
+	gadgetHandler:RegisterAllowCommand(CMD.STOP)
+end
+
+function gadget:GameFrame(frame)
+	local tagUpdates = 0
+	local cmdCache = {}
+
+	for nanoID, targets in pairs(trackingTable) do
+		if nanoID % updateInterval == frame % updateInterval then
+			if next(targets) == nil then
+				trackingTable[nanoID] = nil
+			end
+
+			local nanoDefID = spGetUnitDefID(nanoID)
+			local maxDistance = constructionTurretsDefs[nanoDefID].maxBuildDistance
+
+			for targetID, commandData in pairs(targets) do
+				if spValidUnitID(targetID) then
+					if commandData.cmdTag == -1 then
```

Comment:
`nil` is already the magic value for "no valid value", perhaps use that instead of `-1`.